File: payment-core/go-services/internal/banking/nibss_service.go
package banking

import (
	"bytes"
	"crypto/rand"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"os"
	"regexp"
	"strings"
	"sync"
	"time"
)

type BankAccount struct {
	AccountNumber string `json:"accountNumber"`
	AccountName   string `json:"accountName"`
	BankCode      string `json:"bankCode"`
	BankName      string `json:"bankName"`
	BVN           string `json:"bvn,omitempty"`
	Currency      string `json:"currency"`
}

type BankTransferRequest struct {
	FromAccount string  `json:"fromAccount"`
	ToAccount   string  `json:"toAccount"`
	ToBankCode  string  `json:"toBankCode"`
	Amount      float64 `json:"amount"`
	Narration   string  `json:"narration"`
	Reference   string  `json:"reference"`
}

type BankTransferResponse struct {
	SessionID       string    `json:"sessionId"`
	Reference       string    `json:"reference"`
	ResponseCode    string    `json:"responseCode"`
	ResponseMessage string    `json:"responseMessage"`
	Amount          float64   `json:"amount"`
	TransactionDate time.Time `json:"transactionDate"`
}

type TransferStatus string

const (
	TransferStatusPending    TransferStatus = "pending"
	TransferStatusProcessing TransferStatus = "processing"
	TransferStatusCompleted  TransferStatus = "completed"
	TransferStatusFailed     TransferStatus = "failed"
	TransferStatusReversed   TransferStatus = "reversed"
)

type BankTransferStatus struct {
	Reference       string         `json:"reference"`
	Status          TransferStatus `json:"status"`
	ResponseCode    string         `json:"responseCode"`
	ResponseMessage string         `json:"responseMessage"`
	Amount          float64        `json:"amount,omitempty"`
	CompletedAt     *time.Time     `json:"completedAt,omitempty"`
}

type BVNVerificationResult struct {
	BVN         string  `json:"bvn"`
	FirstName   string  `json:"firstName"`
	LastName    string  `json:"lastName"`
	DateOfBirth string  `json:"dateOfBirth"`
	PhoneNumber string  `json:"phoneNumber"`
	Verified    bool    `json:"verified"`
	MatchScore  float64 `json:"matchScore,omitempty"`
}

type NigerianBank struct {
	Code      string `json:"code"`
	Name      string `json:"name"`
	ShortName string `json:"shortName"`
}

type NIBSSService struct {
	mu              sync.RWMutex
	apiURL          string
	apiKey          string
	institutionCode string
	httpClient      *http.Client
	banks           []NigerianBank
}

func NewNIBSSService() *NIBSSService {
	apiURL := os.Getenv("NIBSS_API_URL")
	if apiURL == "" {
		apiURL = "https://api.nibss-plc.com.ng"
	}

	service := &NIBSSService{
		apiURL:          apiURL,
		apiKey:          os.Getenv("NIBSS_API_KEY"),
		institutionCode: os.Getenv("NIBSS_INSTITUTION_CODE"),
		httpClient: &http.Client{
			Timeout: 30 * time.Second,
		},
		banks: initializeNigerianBanks(),
	}

	return service
}

func initializeNigerianBanks() []NigerianBank {
	return []NigerianBank{
		{Code: "044", Name: "Access Bank Plc", ShortName: "Access Bank"},
		{Code: "063", Name: "Access Bank (Diamond) Plc", ShortName: "Diamond Bank"},
		{Code: "050", Name: "Ecobank Nigeria Plc", ShortName: "Ecobank"},
		{Code: "070", Name: "Fidelity Bank Plc", ShortName: "Fidelity Bank"},
		{Code: "011", Name: "First Bank of Nigeria Limited", ShortName: "First Bank"},
		{Code: "214", Name: "First City Monument Bank Plc", ShortName: "FCMB"},
		{Code: "058", Name: "Guaranty Trust Bank Plc", ShortName: "GTBank"},
		{Code: "030", Name: "Heritage Banking Company Ltd", ShortName: "Heritage Bank"},
		{Code: "301", Name: "Jaiz Bank Plc", ShortName: "Jaiz Bank"},
		{Code: "082", Name: "Keystone Bank Limited", ShortName: "Keystone Bank"},
		{Code: "526", Name: "Parallex Bank Ltd", ShortName: "Parallex Bank"},
		{Code: "076", Name: "Polaris Bank Limited", ShortName: "Polaris Bank"},
		{Code: "101", Name: "Providus Bank", ShortName: "Providus Bank"},
		{Code: "221", Name: "Stanbic IBTC Bank Plc", ShortName: "Stanbic IBTC"},
		{Code: "068", Name: "Standard Chartered Bank Nigeria Ltd", ShortName: "Standard Chartered"},
		{Code: "232", Name: "Sterling Bank Plc", ShortName: "Sterling Bank"},
		{Code: "100", Name: "Suntrust Bank Nigeria Limited", ShortName: "Suntrust Bank"},
		{Code: "032", Name: "Union Bank of Nigeria Plc", ShortName: "Union Bank"},
		{Code: "033", Name: "United Bank For Africa Plc", ShortName: "UBA"},
		{Code: "215", Name: "Unity Bank Plc", ShortName: "Unity Bank"},
		{Code: "035", Name: "Wema Bank Plc", ShortName: "Wema Bank"},
		{Code: "057", Name: "Zenith Bank Plc", ShortName: "Zenith Bank"},
		{Code: "304", Name: "Globus Bank Limited", ShortName: "Globus Bank"},
		{Code: "090175", Name: "Rubies MFB", ShortName: "Rubies MFB"},
		{Code: "090267", Name: "Kuda Bank", ShortName: "Kuda"},
	}
}

func (s *NIBSSService) VerifyBankAccount(accountNumber, bankCode string) (*BankAccount, error) {
	requestID := generateUUID()

	reqBody := map[string]interface{}{
		"requestId":       requestID,
		"accountNumber":   accountNumber,
		"bankCode":        bankCode,
		"institutionCode": s.institutionCode,
	}

	jsonBody, err := json.Marshal(reqBody)
	if err != nil {
		return nil, fmt.Errorf("failed to marshal request: %w", err)
	}

	req, err := http.NewRequest("POST", s.apiURL+"/nameenquiry", bytes.NewBuffer(jsonBody))
	if err != nil {
		return nil, fmt.Errorf("failed to create request: %w", err)
	}

	req.Header.Set("Content-Type", "application/json")
	req.Header.Set("Authorization", "Bearer "+s.apiKey)

	resp, err := s.httpClient.Do(req)
	if err != nil {
		return nil, fmt.Errorf("NIBSS API request failed: %w", err)
	}
	defer resp.Body.Close()

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, fmt.Errorf("failed to read response: %w", err)
	}

	if resp.StatusCode != http.StatusOK {
		return nil, fmt.Errorf("NIBSS verification error: %s", string(body))
	}

	var data map[string]interface{}
	if err := json.Unmarshal(body, &data); err != nil {
		return nil, fmt.Errorf("failed to parse response: %w", err)
	}

	responseCode, _ := data["responseCode"].(string)
	if responseCode != "00" {
		responseMessage, _ := data["responseMessage"].(string)
		return nil, fmt.Errorf("account verification failed: %s", responseMessage)
	}

	accountName, _ := data["accountName"].(string)
	bvn, _ := data["bvn"].(string)

	return &BankAccount{
		AccountNumber: accountNumber,
		AccountName:   accountName,
		BankCode:      bankCode,
		BankName:      s.GetBankName(bankCode),
		BVN:           bvn,
		Currency:      "NGN",
	}, nil
}

func (s *NIBSSService) InitiateTransfer(req *BankTransferRequest) (*BankTransferResponse, error) {
	sessionID := generateUUID()

	reqBody := map[string]interface{}{
		"sessionId":        sessionID,
		"fromAccount":      req.FromAccount,
		"toAccount":        req.ToAccount,
		"toBankCode":       req.ToBankCode,
		"amount":           req.Amount,
		"narration":        req.Narration,
		"paymentReference": req.Reference,
		"institutionCode":  s.institutionCode,
		"channelCode":      "7",
	}

	jsonBody, err := json.Marshal(reqBody)
	if err != nil {
		return nil, fmt.Errorf("failed to marshal request: %w", err)
	}

	httpReq, err := http.NewRequest("POST", s.apiURL+"/nip/fundstransfer", bytes.NewBuffer(jsonBody))
	if err != nil {
		return nil, fmt.Errorf("failed to create request: %w", err)
	}

	httpReq.Header.Set("Content-Type", "application/json")
	httpReq.Header.Set("Authorization", "Bearer "+s.apiKey)

	resp, err := s.httpClient.Do(httpReq)
	if err != nil {
		return nil, fmt.Errorf("NIBSS API request failed: %w", err)
	}
	defer resp.Body.Close()

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, fmt.Errorf("failed to read response: %w", err)
	}

	if resp.StatusCode != http.StatusOK {
		return nil, fmt.Errorf("NIBSS transfer error: %s", string(body))
	}

	var data map[string]interface{}
	if err := json.Unmarshal(body, &data); err != nil {
		return nil, fmt.Errorf("failed to parse response: %w", err)
	}

	respSessionID, _ := data["sessionId"].(string)
	if respSessionID == "" {
		respSessionID = sessionID
	}

	responseCode, _ := data["responseCode"].(string)
	responseMessage, _ := data["responseMessage"].(string)
	transactionDateStr, _ := data["transactionDate"].(string)

	transactionDate := time.Now()
	if transactionDateStr != "" {
		if parsed, err := time.Parse(time.RFC3339, transactionDateStr); err == nil {
			transactionDate = parsed
		}
	}

	return &BankTransferResponse{
		SessionID:       respSessionID,
		Reference:       req.Reference,
		ResponseCode:    responseCode,
		ResponseMessage: responseMessage,
		Amount:          req.Amount,
		TransactionDate: transactionDate,
	}, nil
}

func (s *NIBSSService) GetTransferStatus(reference, sessionID string) (*BankTransferStatus, error) {
	reqBody := map[string]interface{}{
		"sessionId":        sessionID,
		"paymentReference": reference,
		"institutionCode":  s.institutionCode,
	}

	jsonBody, err := json.Marshal(reqBody)
	if err != nil {
		return nil, fmt.Errorf("failed to marshal request: %w", err)
	}

	req, err := http.NewRequest("POST", s.apiURL+"/nip/transactionstatus", bytes.NewBuffer(jsonBody))
	if err != nil {
		return nil, fmt.Errorf("failed to create request: %w", err)
	}

	req.Header.Set("Content-Type", "application/json")
	req.Header.Set("Authorization", "Bearer "+s.apiKey)

	resp, err := s.httpClient.Do(req)
	if err != nil {
		return nil, fmt.Errorf("NIBSS API request failed: %w", err)
	}
	defer resp.Body.Close()

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, fmt.Errorf("failed to read response: %w", err)
	}

	if resp.StatusCode != http.StatusOK {
		return nil, fmt.Errorf("failed to get transfer status: %s", string(body))
	}

	var data map[string]interface{}
	if err := json.Unmarshal(body, &data); err != nil {
		return nil, fmt.Errorf("failed to parse response: %w", err)
	}

	responseCode, _ := data["responseCode"].(string)
	responseMessage, _ := data["responseMessage"].(string)
	amount, _ := data["amount"].(float64)

	var status TransferStatus
	switch responseCode {
	case "00":
		status = TransferStatusCompleted
	case "09":
		status = TransferStatusProcessing
	case "51":
		status = TransferStatusFailed
	case "56":
		status = TransferStatusReversed
	default:
		status = TransferStatusPending
	}

	result := &BankTransferStatus{
		Reference:       reference,
		Status:          status,
		ResponseCode:    responseCode,
		ResponseMessage: responseMessage,
		Amount:          amount,
	}

	if status == TransferStatusCompleted {
		now := time.Now()
		result.CompletedAt = &now
	}

	return result, nil
}

func (s *NIBSSService) VerifyBVN(bvn, firstName, lastName, dateOfBirth string) (*BVNVerificationResult, error) {
	requestID := generateUUID()

	reqBody := map[string]interface{}{
		"requestId":       requestID,
		"bvn":             bvn,
		"firstName":       firstName,
		"lastName":        lastName,
		"dateOfBirth":     dateOfBirth,
		"institutionCode": s.institutionCode,
	}

	jsonBody, err := json.Marshal(reqBody)
	if err != nil {
		return nil, fmt.Errorf("failed to marshal request: %w", err)
	}

	req, err := http.NewRequest("POST", s.apiURL+"/bvnverification", bytes.NewBuffer(jsonBody))
	if err != nil {
		return nil, fmt.Errorf("failed to create request: %w", err)
	}

	req.Header.Set("Content-Type", "application/json")
	req.Header.Set("Authorization", "Bearer "+s.apiKey)

	resp, err := s.httpClient.Do(req)
	if err != nil {
		return nil, fmt.Errorf("NIBSS API request failed: %w", err)
	}
	defer resp.Body.Close()

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, fmt.Errorf("failed to read response: %w", err)
	}

	if resp.StatusCode != http.StatusOK {
		return nil, fmt.Errorf("BVN verification error: %s", string(body))
	}

	var data map[string]interface{}
	if err := json.Unmarshal(body, &data); err != nil {
		return nil, fmt.Errorf("failed to parse response: %w", err)
	}

	responseCode, _ := data["responseCode"].(string)
	if responseCode != "00" {
		responseMessage, _ := data["responseMessage"].(string)
		return nil, fmt.Errorf("BVN verification failed: %s", responseMessage)
	}

	respFirstName, _ := data["firstName"].(string)
	respLastName, _ := data["lastName"].(string)
	respDOB, _ := data["dateOfBirth"].(string)
	phoneNumber, _ := data["phoneNumber"].(string)

	var matchScore float64
	if firstName != "" || lastName != "" || dateOfBirth != "" {
		matches := 0
		total := 0

		if firstName != "" {
			total++
			if strings.EqualFold(respFirstName, firstName) {
				matches++
			}
		}
		if lastName != "" {
			total++
			if strings.EqualFold(respLastName, lastName) {
				matches++
			}
		}
		if dateOfBirth != "" {
			total++
			if respDOB == dateOfBirth {
				matches++
			}
		}

		if total > 0 {
			matchScore = float64(matches) / float64(total) * 100
		}
	}

	return &BVNVerificationResult{
		BVN:         bvn,
		FirstName:   respFirstName,
		LastName:    respLastName,
		DateOfBirth: respDOB,
		PhoneNumber: phoneNumber,
		Verified:    responseCode == "00",
		MatchScore:  matchScore,
	}, nil
}

func (s *NIBSSService) GetNigerianBanks() []NigerianBank {
	s.mu.RLock()
	defer s.mu.RUnlock()
	return s.banks
}

func (s *NIBSSService) GetBankName(code string) string {
	s.mu.RLock()
	defer s.mu.RUnlock()

	for _, bank := range s.banks {
		if bank.Code == code {
			return bank.Name
		}
	}
	return "Unknown Bank"
}

func (s *NIBSSService) GetBankCode(name string) string {
	s.mu.RLock()
	defer s.mu.RUnlock()

	nameLower := strings.ToLower(name)
	for _, bank := range s.banks {
		if strings.Contains(strings.ToLower(bank.Name), nameLower) ||
			strings.Contains(strings.ToLower(bank.ShortName), nameLower) {
			return bank.Code
		}
	}
	return ""
}

func ValidateAccountNumber(accountNumber string) bool {
	matched, _ := regexp.MatchString(`^\d{10}$`, accountNumber)
	return matched
}

func ValidateBVN(bvn string) bool {
	matched, _ := regexp.MatchString(`^\d{11}$`, bvn)
	return matched
}

func GenerateTransferReference(prefix string) string {
	if prefix == "" {
		prefix = "REM"
	}
	timestamp := time.Now().UnixMilli()
	randomBytes := make([]byte, 4)
	rand.Read(randomBytes)
	random := strings.ToUpper(hex.EncodeToString(randomBytes))
	return fmt.Sprintf("%s%X%s", prefix, timestamp, random)
}

func CalculateTransferFee(amount float64) float64 {
	if amount <= 5000 {
		return 10
	} else if amount <= 50000 {
		return 25
	}
	return 50
}

func ValidateTransferAmount(amount float64) (bool, string) {
	const minAmount = 100
	const maxAmount = 10000000

	if amount < minAmount {
		return false, fmt.Sprintf("Minimum transfer amount is NGN %s", formatCurrency(minAmount))
	}
	if amount > maxAmount {
		return false, fmt.Sprintf("Maximum transfer amount is NGN %s", formatCurrency(maxAmount))
	}
	return true, ""
}

func FormatAmountForNIBSS(amountInNaira float64) int64 {
	return int64(amountInNaira * 100)
}

func ParseAmountFromNIBSS(amountInKobo int64) float64 {
	return float64(amountInKobo) / 100
}

func (s *NIBSSService) RetryTransfer(req *BankTransferRequest, maxRetries int) (*BankTransferResponse, error) {
	if maxRetries <= 0 {
		maxRetries = 3
	}

	var lastErr error
	for attempt := 1; attempt <= maxRetries; attempt++ {
		resp, err := s.InitiateTransfer(req)
		if err == nil {
			return resp, nil
		}

		lastErr = err

		if strings.Contains(err.Error(), "invalid") {
			return nil, err
		}

		if attempt < maxRetries {
			delay := time.Duration(1<<uint(attempt)) * time.Second
			time.Sleep(delay)
		}
	}

	return nil, lastErr
}

func generateUUID() string {
	b := make([]byte, 16)
	rand.Read(b)
	return fmt.Sprintf("%x-%x-%x-%x-%x", b[0:4], b[4:6], b[6:8], b[8:10], b[10:])
}

func formatCurrency(amount float64) string {
	return fmt.Sprintf("%.2f", amount)
}

File: payment-switch/server/services/nibssService.ts
/**
 * NIBSS (Nigeria Inter-Bank Settlement System) Integration Service
 * 
 * Handles Nigerian bank account verification and instant payments via NIP
 * Supports 25+ Nigerian commercial banks
 */

import crypto from 'crypto';

// NIBSS API configuration
const NIBSS_API_URL = process.env.NIBSS_API_URL || 'https://api.nibss-plc.com.ng';
const NIBSS_API_KEY = process.env.NIBSS_API_KEY || '';
const NIBSS_INSTITUTION_CODE = process.env.NIBSS_INSTITUTION_CODE || '';

export interface BankAccount {
  accountNumber: string;
  accountName: string;
  bankCode: string;
  bankName: string;
  bvn?: string;
  currency: string;
}

export interface BankTransferRequest {
  fromAccount: string;
  toAccount: string;
  toBankCode: string;
  amount: number;
  narration: string;
  reference: string;
}

export interface BankTransferResponse {
  sessionId: string;
  reference: string;
  responseCode: string;
  responseMessage: string;
  amount: number;
  transactionDate: string;
}

export interface BankTransferStatus {
  reference: string;
  status: 'pending' | 'processing' | 'completed' | 'failed' | 'reversed';
  responseCode: string;
  responseMessage: string;
  amount?: number;
  completedAt?: Date;
}

/**
 * Verify bank account details via NIBSS Name Enquiry Service
 */
export async function verifyBankAccount(params: {
  accountNumber: string;
  bankCode: string;
}): Promise<BankAccount> {
  const requestId = crypto.randomUUID();
  
  const response = await fetch(`${NIBSS_API_URL}/nameenquiry`, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
      'Authorization': `Bearer ${NIBSS_API_KEY}`,
    },
    body: JSON.stringify({
      requestId,
      accountNumber: params.accountNumber,
      bankCode: params.bankCode,
      institutionCode: NIBSS_INSTITUTION_CODE,
    }),
  });

  if (!response.ok) {
    const error = await response.json();
    throw new Error(`NIBSS verification error: ${error.message || response.statusText}`);
  }

  const data = await response.json();

  if (data.responseCode !== '00') {
    throw new Error(`Account verification failed: ${data.responseMessage}`);
  }

  return {
    accountNumber: params.accountNumber,
    accountName: data.accountName,
    bankCode: params.bankCode,
    bankName: data.bankName || getNigerianBankName(params.bankCode),
    bvn: data.bvn,
    currency: 'NGN',
  };
}

/**
 * Initiate instant bank transfer via NIBSS Instant Payment (NIP)
 */
export async function initiateTransfer(params: BankTransferRequest): Promise<BankTransferResponse> {
  const sessionId = crypto.randomUUID();
  
  const response = await fetch(`${NIBSS_API_URL}/nip/fundstransfer`, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
      'Authorization': `Bearer ${NIBSS_API_KEY}`,
    },
    body: JSON.stringify({
      sessionId,
      fromAccount: params.fromAccount,
      toAccount: params.toAccount,
      toBankCode: params.toBankCode,
      amount: params.amount,
      narration: params.narration,
      paymentReference: params.reference,
      institutionCode: NIBSS_INSTITUTION_CODE,
      channelCode: '7', // Internet banking
    }),
  });

  if (!response.ok) {
    const error = await response.json();
    throw new Error(`NIBSS transfer error: ${error.message || response.statusText}`);
  }

  const data = await response.json();

  return {
    sessionId: data.sessionId || sessionId,
    reference: params.reference,
    responseCode: data.responseCode,
    responseMessage: data.responseMessage,
    amount: params.amount,
    transactionDate: data.transactionDate || new Date().toISOString(),
  };
}

/**
 * Query transfer status
 */
export async function getTransferStatus(params: {
  reference: string;
  sessionId: string;
}): Promise<BankTransferStatus> {
  const response = await fetch(`${NIBSS_API_URL}/nip/transactionstatus`, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
      'Authorization': `Bearer ${NIBSS_API_KEY}`,
    },
    body: JSON.stringify({
      sessionId: params.sessionId,
      paymentReference: params.reference,
      institutionCode: NIBSS_INSTITUTION_CODE,
    }),
  });

  if (!response.ok) {
    throw new Error(`Failed to get transfer status: ${response.statusText}`);
  }

  const data = await response.json();

  // Map NIBSS response codes to status
  let status: BankTransferStatus['status'] = 'pending';
  
  if (data.responseCode === '00') {
    status = 'completed';
  } else if (data.responseCode === '09') {
    status = 'processing';
  } else if (data.responseCode === '51') {
    status = 'failed';
  } else if (data.responseCode === '56') {
    status = 'reversed';
  }

  return {
    reference: params.reference,
    status,
    responseCode: data.responseCode,
    responseMessage: data.responseMessage,
    amount: data.amount,
    completedAt: status === 'completed' ? new Date(data.transactionDate) : undefined,
  };
}

/**
 * Verify BVN (Bank Verification Number)
 */
export async function verifyBVN(params: {
  bvn: string;
  firstName?: string;
  lastName?: string;
  dateOfBirth?: string; // YYYY-MM-DD
}): Promise<{
  bvn: string;
  firstName: string;
  lastName: string;
  dateOfBirth: string;
  phoneNumber: string;
  verified: boolean;
  matchScore?: number;
}> {
  const requestId = crypto.randomUUID();

  const response = await fetch(`${NIBSS_API_URL}/bvnverification`, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
      'Authorization': `Bearer ${NIBSS_API_KEY}`,
    },
    body: JSON.stringify({
      requestId,
      bvn: params.bvn,
      firstName: params.firstName,
      lastName: params.lastName,
      dateOfBirth: params.dateOfBirth,
      institutionCode: NIBSS_INSTITUTION_CODE,
    }),
  });

  if (!response.ok) {
    const error = await response.json();
    throw new Error(`BVN verification error: ${error.message || response.statusText}`);
  }

  const data = await response.json();

  if (data.responseCode !== '00') {
    throw new Error(`BVN verification failed: ${data.responseMessage}`);
  }

  // Calculate match score if comparison data was provided
  let matchScore: number | undefined;
  if (params.firstName || params.lastName || params.dateOfBirth) {
    let matches = 0;
    let total = 0;

    if (params.firstName) {
      total++;
      if (data.firstName?.toLowerCase() === params.firstName.toLowerCase()) matches++;
    }
    if (params.lastName) {
      total++;
      if (data.lastName?.toLowerCase() === params.lastName.toLowerCase()) matches++;
    }
    if (params.dateOfBirth) {
      total++;
      if (data.dateOfBirth === params.dateOfBirth) matches++;
    }

    matchScore = total > 0 ? (matches / total) * 100 : undefined;
  }

  return {
    bvn: params.bvn,
    firstName: data.firstName,
    lastName: data.lastName,
    dateOfBirth: data.dateOfBirth,
    phoneNumber: data.phoneNumber,
    verified: data.responseCode === '00',
    matchScore,
  };
}

/**
 * Get list of Nigerian banks with NIBSS codes
 */
export function getNigerianBanks(): Array<{
  code: string;
  name: string;
  shortName: string;
}> {
  return [
    { code: '044', name: 'Access Bank Plc', shortName: 'Access Bank' },
    { code: '063', name: 'Access Bank (Diamond) Plc', shortName: 'Diamond Bank' },
    { code: '050', name: 'Ecobank Nigeria Plc', shortName: 'Ecobank' },
    { code: '070', name: 'Fidelity Bank Plc', shortName: 'Fidelity Bank' },
    { code: '011', name: 'First Bank of Nigeria Limited', shortName: 'First Bank' },
    { code: '214', name: 'First City Monument Bank Plc', shortName: 'FCMB' },
    { code: '058', name: 'Guaranty Trust Bank Plc', shortName: 'GTBank' },
    { code: '030', name: 'Heritage Banking Company Ltd', shortName: 'Heritage Bank' },
    { code: '301', name: 'Jaiz Bank Plc', shortName: 'Jaiz Bank' },
    { code: '082', name: 'Keystone Bank Limited', shortName: 'Keystone Bank' },
    { code: '526', name: 'Parallex Bank Ltd', shortName: 'Parallex Bank' },
    { code: '076', name: 'Polaris Bank Limited', shortName: 'Polaris Bank' },
    { code: '101', name: 'Providus Bank', shortName: 'Providus Bank' },
    { code: '221', name: 'Stanbic IBTC Bank Plc', shortName: 'Stanbic IBTC' },
    { code: '068', name: 'Standard Chartered Bank Nigeria Ltd', shortName: 'Standard Chartered' },
    { code: '232', name: 'Sterling Bank Plc', shortName: 'Sterling Bank' },
    { code: '100', name: 'Suntrust Bank Nigeria Limited', shortName: 'Suntrust Bank' },
    { code: '032', name: 'Union Bank of Nigeria Plc', shortName: 'Union Bank' },
    { code: '033', name: 'United Bank For Africa Plc', shortName: 'UBA' },
    { code: '215', name: 'Unity Bank Plc', shortName: 'Unity Bank' },
    { code: '035', name: 'Wema Bank Plc', shortName: 'Wema Bank' },
    { code: '057', name: 'Zenith Bank Plc', shortName: 'Zenith Bank' },
    { code: '304', name: 'Globus Bank Limited', shortName: 'Globus Bank' },
    { code: '090175', name: 'Rubies MFB', shortName: 'Rubies MFB' },
    { code: '090267', name: 'Kuda Bank', shortName: 'Kuda' },
  ];
}

/**
 * Get bank name from code
 */
export function getNigerianBankName(code: string): string {
  const bank = getNigerianBanks().find(b => b.code === code);
  return bank?.name || 'Unknown Bank';
}

/**
 * Get bank code from name
 */
export function getNigerianBankCode(name: string): string | undefined {
  const bank = getNigerianBanks().find(
    b => b.name.toLowerCase().includes(name.toLowerCase()) ||
         b.shortName.toLowerCase().includes(name.toLowerCase())
  );
  return bank?.code;
}

/**
 * Validate Nigerian account number format
 */
export function validateAccountNumber(accountNumber: string): boolean {
  // Nigerian account numbers are typically 10 digits
  return /^\d{10}$/.test(accountNumber);
}

/**
 * Validate BVN format
 */
export function validateBVN(bvn: string): boolean {
  // BVN is 11 digits
  return /^\d{11}$/.test(bvn);
}

/**
 * Generate transfer reference
 */
export function generateTransferReference(prefix: string = 'REM'): string {
  const timestamp = Date.now().toString(36).toUpperCase();
  const random = crypto.randomBytes(4).toString('hex').toUpperCase();
  return `${prefix}${timestamp}${random}`;
}

/**
 * Calculate NIBSS transfer fee
 */
export function calculateTransferFee(amount: number): number {
  // NIBSS NIP fees (as of 2024)
  if (amount <= 5000) {
    return 10; // ₦10
  } else if (amount <= 50000) {
    return 25; // ₦25
  } else {
    return 50; // ₦50
  }
}

/**
 * Check if transfer amount is within limits
 */
export function validateTransferAmount(amount: number): {
  valid: boolean;
  error?: string;
} {
  const MIN_AMOUNT = 100; // ₦100
  const MAX_AMOUNT = 10000000; // ₦10,000,000 (10 million)

  if (amount < MIN_AMOUNT) {
    return {
      valid: false,
      error: `Minimum transfer amount is ₦${MIN_AMOUNT.toLocaleString()}`,
    };
  }

  if (amount > MAX_AMOUNT) {
    return {
      valid: false,
      error: `Maximum transfer amount is ₦${MAX_AMOUNT.toLocaleString()}`,
    };
  }

  return { valid: true };
}

/**
 * Format amount for NIBSS API (kobo to naira)
 */
export function formatAmountForNIBSS(amountInNaira: number): number {
  // NIBSS expects amount in kobo (1 naira = 100 kobo)
  return Math.round(amountInNaira * 100);
}

/**
 * Parse amount from NIBSS API (kobo to naira)
 */
export function parseAmountFromNIBSS(amountInKobo: number): number {
  return amountInKobo / 100;
}

/**
 * Retry transfer with exponential backoff
 */
export async function retryTransfer(
  params: BankTransferRequest,
  maxRetries: number = 3
): Promise<BankTransferResponse> {
  let lastError: Error;

  for (let attempt = 1; attempt <= maxRetries; attempt++) {
    try {
      return await initiateTransfer(params);
    } catch (error) {
      lastError = error as Error;

      // Don't retry on client errors (invalid account, insufficient funds, etc.)
      if (error instanceof Error && error.message.includes('invalid')) {
        throw error;
      }

      // Wait before retrying (exponential backoff)
      if (attempt < maxRetries) {
        const delay = Math.pow(2, attempt) * 1000; // 2s, 4s, 8s
        await new Promise(resolve => setTimeout(resolve, delay));
      }
    }
  }

  throw lastError!;
}
